### DISCUSSION TITLE
Move to HA translation strings

### DISCUSSION THREAD

**Discussion by 13robin37** [UI-Lovelace-Minimalist/UI#711] [Ideas]
I've tested with the hass object a bit, turns out you can fetch the translations from Home Assistant directly.

```
                    let state = entity.state;
                    let lang = hass["language"];
                    label = hass.resources[lang]["component.vacuum.state._." + state];
                    return label;
```
Just an example to get the translation for the state based on the vacuum component. This fetches the current device state, detects the user language and returns the localized string.
Would be a cool feature to move to this in the long run I think. That way we would basically hook onto the way more extensive translations from HA and not worry about languages that much anymore.

Thoughts?

**Comment by basbruss** [comment]
This could be something for a `v1.0.0` release. 🎉 
Are all strings we translate included in the HA database? Haven't looked at it yet.
And do you prefer it to only feature on official cards or also on custom cards?

**Reply by 13robin37** [reply]
Yes, they should be present. Currently testing the best way to provide the translated string automatically based on the entity provided, once I have a more detailed overview I will suggest some changes. Shouldn't be breaking hopefully. 🙏 
I would use this for all cards, maybe with a variable like ulm_state to allow easy implementation. We'll see!

**Reply by basbruss** [reply]
I have tried some strings based on the states defined in HA https://developers.home-assistant.io/docs/core/entity/.
And it works like magic! Do you want to start a PR to add this new form of translation? 

I noticed some device entities did not give any results like `alarm_control_panel`.

Edit: @13robin37 I have altered the template slightly to also auto-detect the type of the entity. And added a fallback when no translations are available. For example manual configured zones.
```java
  let state = entity.state;
  let lang = hass["language"]
  let type = entity.entity_id.substr(0, entity.entity_id.indexOf("."));
  let label = hass.resources[lang]["component." + type + ".state._." + state] || entity.state;
  return label
```

**Reply by 13robin37** [reply]
Strange, you can check the translation/entity list yourself with `console.log(hass.resources);` which outputs multiple translations regarding the `alarm_control_panel`. Personally I don't use that one yet. Will test with my available entities and do a PR once I finished testing.

**Reply by basbruss** [reply]
@13robin37 I have found out why it was not working. Not all strings are reachable with `component.type.state._.state`.
The alarm panel use for instance `state_badge.alarm_control_panel.armed` and binary_sensors use `component.binary_sensor.state.device_class.state`

**Reply by 13robin37** [reply]
I will try to build a condition for that to fetch states from the most common entities. That way all basic cases should be covered.

**Reply by basbruss** [reply]
I have done some more testing and have created a workable structure 😄 .

> I noticed some device entities did not give any results like `alarm_control_panel`

This was probably due to some wrong string conventions with testing. As it works now.
Also checking the console log does not show entries for all types of entities however it does work.

Maybe adding the translation template in a variable to make it easily accessible is a good approach.

```yaml
ulm_translation_engine:
  variables:
    ulm_translation: >
      [[[
        let state = entity.state;
        let d_class = entity.attributes.device_class;
        let def = ["unknown", "unavailable"];
        let lang = hass["language"]
        let domain = entity.entity_id.substr(0, entity.entity_id.indexOf("."));
        var translation = hass.resources[lang]["component." + domain + ".state._." + state] || entity.state;
        if (def.includes(state)) {
          var translation = hass.resources[lang]["state.default." + state ]
        }
        else if (domain == "binary_sensor" && d_class != ''){
          var translation = hass.resources[lang]["component." + domain + ".state." + d_class + "." + state] || entity.state;
        }
        return translation;
      ]]]
```
The states `["unknown", "unavailable"];` are not accessible through the component but as `state.default`
`binary_sensors` offer better translations through their device_class.

Edit: Tested in on the label in the light card and it seems to work perfect this way
```yaml
label: >-
  [[[
    var label = variables.ulm_translation
    if (entity.state == "on" && entity.attributes.brightness != null){
      var bri = Math.round(entity.attributes.brightness / 2.55);
      var label = (bri ? bri : "0") + "%";
    }
    return label
  ]]]
```
Hope this helps you a bit!

**Reply by 13robin37** [reply]
Awesome, thanks for your very generous support! Will look into this once I have free time (and some non-busy brain cells).
